R-package/tests/testthat/test_model_compatibility.R

```diff
@@ -72,13 +72,34 @@ test_that("Models from previous versions of XGBoost can be loaded", {
     m <- regmatches(model_file, m)[[1]]
     model_xgb_ver <- m[2]
     name <- m[3]
+    is_rds <- endsWith(model_file, '.rds')
 
-    if (endsWith(model_file, '.rds')) {
-      booster <- readRDS(model_file)
-    } else {
-      booster <- xgb.load(model_file)
+    cpp_warning <- capture.output({
+      # Expect an R warning when a model is loaded from RDS and it was generated by version < 1.1.x
+      if (is_rds && compareVersion(model_xgb_ver, '1.1.1.1') < 0) {
```

Comment:
I am a bit confused by this logic. So for ver >= 1.1.1.1, `predict` and `run_booster_check` no longer throw warnings when the model is loaded from RDS? Wouldn't you want to discourage it regardless and throw warning?

Reply:
That's correct. When RDS is used, XGBoost package does not actually know that RDS is used. So the warning is only generated when a material change in the model representation is detected. There was a major change in 1.1.1.1, and a minor one in 1.2.0.1. The major change is covered by the R warning ("The model had been generated by XGBoost version 1.0.0 or earlier and was loaded from a RDS file") and the minor change is covered by the C++ warning ("Attempted to load internal configuration for a model file that was generated by a previous version of XGBoost").

Reply:
The official documentation has a clear guidance against RDS: https://www.rdocumentation.org/packages/xgboost/versions/1.2.0.1/topics/a-compatibility-note-for-saveRDS-save